app/services/otp_rate_limiter.rb

```diff
@@ -0,0 +1,65 @@
+class OtpRateLimiter
+  def initialize(phone:, user:)
+    @phone = phone
+    @user = user
+  end
+
+  def exceeded_otp_send_limit?
+    return false if entry_for_current_phone.blank?
+
+    if rate_limit_period_expired?
+      reset_count_and_otp_last_sent_at
+      return false
+    end
+
+    max_requests_reached?
+  end
+
+  def max_requests_reached?
+    entry_for_current_phone.otp_send_count >= otp_maxretry_times
+  end
+
+  def rate_limit_period_expired?
+    otp_last_sent_at.present? && (otp_last_sent_at + otp_findtime) < Time.zone.now
+  end
+
+  def reset_count_and_otp_last_sent_at
+    entry_for_current_phone.update(otp_last_sent_at: Time.zone.now, otp_send_count: 0)
+  end
+
+  def lock_out_user
+    UpdateUser.new(user: user, attributes: { second_factor_locked_at: Time.zone.now }).call
+  end
+
+  def increment
+    now = Time.zone.now
+
+    if entry_for_current_phone.present?
+      entry_for_current_phone.otp_send_count += 1
+      entry_for_current_phone.otp_last_sent_at = now
+      entry_for_current_phone.save!
```

Comment:
So I think (like you mentioned) there is a small opportunity for a race condition here.

I think that if possible we should take advantage of Rail's `first_or_create` (as well as a top-level retry), so something like:

```ruby
def find_or_create_with_phone(phone)
  return if phone.blank? # maybe just raise an error here?
  begin
    OtpRequestsTracker.where(phone_fingerprint: Pii::Fingerprinter.fingerprint(phone)).
      first_or_create(otp_send_count: 0, otp_last_sent_at: Time.zone.now)
  rescue ActiveRecord::RecordNotUnique
    retry # possibly use with_retries gem or add additional logic to only retry once
  end
end
```